python/cudf/cudf/_libxx/table.pyx

```diff
@@ -1,79 +1,144 @@
 
 
-class OrderedColumnDict(OrderedDict):
-    def __setitem__(self, key, value):
-        from cudf.core.column import ColumnBase
+cdef class Table:
 
-        if not isinstance(value, ColumnBase):
-            raise TypeError(
-                f"Cannot insert object of type "
-                "{value.__class__.__name__} into OrderedColumnDict"
-            )
+    def __init__(self, data=None, index=None):
+        """
+        Data: an iterable of Columns
+        """
+        if data is None:
+            data = {}
+        self._data = OrderedColumnDict(data)
+        self._index = index
 
-        if self.first is not None and len(self.first) > 0:
-            if len(value) != len(self.first):
-                raise ValueError(
-                    f"Cannot insert Column of different length "
-                    "into OrderedColumnDict"
-                )
+    @property
+    def _column_names(self):
+        return tuple(self._data.keys())
 
-        super().__setitem__(key, value)
+    @property
+    def _index_names(self):
+        return tuple(self._index._data.keys())
 
     @property
-    def first(self):
+    def _columns(self):
         """
-        Returns the first value if self is non-empty;
-        returns None otherwise.
+        Return a list of Column objects backing this dataframe
         """
-        if len(self) == 0:
-            return None
-        else:
-            return next(iter(self.values()))
-
-
-cdef class _Table:
+        return tuple(self._data.values())
+    
+    @staticmethod
+    cdef Table from_unique_ptr(unique_ptr[table] c_tbl,
+                               column_names,
+                               index_names=None):
+        """
+        Construct a Table from a unique_ptr to a cudf::table.
 
-    def __init__(self, data=None, names=None):
-        if data is None:
-            data = OrderedColumnDict()
-        if isinstance(data, OrderedColumnDict):
-            self._data = data
-        else:
-            if names is None:
-                names = range(len(data))
-            self._data = OrderedColumnDict(zip(names, data))
+        Parameters
+        ----------
+        c_tbl : unique_ptr[cudf::table]
+        index_names : iterable
+        column_names : iterable
+        """
+        cdef vector[unique_ptr[column]] columns
+        columns = c_tbl.get()[0].release()
 
+        cdef vector[unique_ptr[column]].iterator it = columns.begin()
+
+        # First construct the index, if any
+        index = None
+        if index_names is not None:
+            index_columns = []
+            for _ in index_names:
+                index_columns.append(Column.from_unique_ptr(move(dereference(it))))
+                it += 1
+            index = Table(OrderedColumnDict(zip(index_names, index_columns)))
+
+        # Construct the data OrderedColumnDict
+        data_columns = []
+        for _ in column_names:
+            data_columns.append(Column.from_unique_ptr(move(dereference(it))))
```

Comment:
use list comprehension instead of appending

Reply:
I don't like this either, but I need to explicitly increment the iterator `it`. Maybe I'm missing an obvious solution here?

Reply:
Hmm, good point. This is fine then. Normally you avoid doing this because looking up the `append` attribute repeatedly is minimally expensive.